File: packages/mlmpfr.4.0.2+dune/testsuite/exceptionrelatedfunctions.ml
(* This file is part of mlmpfr.

  mlmpfr is free software: you can redistribute it and/or modify it
  under the terms of the GNU Lesser General Public License as
  published by the Free Software Foundation, either version 3 of the
  License, or (at your option) any later version.

  mlmpfr is distributed in the hope that it will be useful, but
  WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
  Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public
  License along with mlmpfr. If not, see
  <http://www.gnu.org/licenses/>. *)

open Printf
module M = Mpfr

(** Convert flags list to MPFR mask *)
let rec flags_to_mask acc f =
  match f with
    [] -> acc
  | h::t ->
     let i = (match h with
              | M.Nan       -> 4
              | M.Divby0    -> 32
              | M.Erange    -> 16
              | M.Underflow -> 1
              | M.Overflow  -> 2
              | M.Inexact   -> 8
              | M.All       -> 63) in
     flags_to_mask (acc+i) t


let flags _ =
  printf "%b %b %b %b %b %b\n"
    (M.underflow_p ())
    (M.overflow_p ())
    (M.divby0_p ())
    (M.nanflag_p ())
    (M.inexflag_p ())
    (M.erangeflag_p ())


let all op em1 em2 =
  printf "%d %d\n" (M.get_emin ()) (M.get_emax ());

  (M.set_emin em1); (M.set_emax em2);
  printf "%d %d\n" (M.get_emin ()) (M.get_emax ());

  printf "%d %d %d %d\n" (M.get_emin_min ()) (M.get_emin_max ()) (M.get_emax_min ()) (M.get_emax_max ());

  printf "%s\n" (M.get_formatted_str (M.check_range op));
  printf "%s\n" (M.get_formatted_str (M.subnormalize op));
  flags ();

  M.set_underflow ();
  M.set_overflow ();
  M.set_divby0 ();
  M.set_nanflag ();
  M.set_inexflag ();
  M.set_erangeflag ();
  flags ();

  M.clear_underflow ();
  M.clear_overflow ();
  M.clear_divby0 ();
  M.clear_nanflag ();
  M.clear_inexflag ();
  M.clear_erangeflag ();
  flags ();

  M.flags_set([M.Nan; M.Erange; M.Divby0]);
  flags ();

  let saved_flags = M.flags_save() in
  let test_all_flags = M.flags_test([M.All]) in
  let test_one_flag = M.flags_test([M.Erange]) in
  printf "%d %d\n" (flags_to_mask 0 test_all_flags) (flags_to_mask 0 test_one_flag);

  M.flags_clear([M.All]);
  flags ();

  M.flags_restore saved_flags [M.Erange; M.Divby0; M.Overflow];
  flags ()

let _ =
  all (M.make_from_float (1. /. 3.)) (-123) 123; printf "\n";
  Gc.full_major (); (* garbage collector full major *)

File: packages/mlmpfr.4.1.0/testsuite/exceptionrelatedfunctions.c
/* This file is part of mlmpfr.

  mlmpfr is free software: you can redistribute it and/or modify it
  under the terms of the GNU Lesser General Public License as
  published by the Free Software Foundation, either version 3 of the
  License, or (at your option) any later version.

  mlmpfr is distributed in the hope that it will be useful, but
  WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
  Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public
  License along with mlmpfr. If not, see
 <http://www.gnu.org/licenses/>. */

#include <stdio.h>
#include <mpfr.h>

void flags ()
{
  printf ("%s %s %s %s %s %s\n",
	  mpfr_underflow_p () ? "true" : "false",
	  mpfr_overflow_p () ? "true" : "false",
	  mpfr_divby0_p () ? "true" : "false",
	  mpfr_nanflag_p () ? "true" : "false",
	  mpfr_inexflag_p () ? "true" : "false",
	  mpfr_erangeflag_p () ? "true" : "false");
}

void all (mpfr_t op, int t, mpfr_exp_t em1, mpfr_exp_t em2)
{
  mpfr_t rop;
  mpfr_init2 (rop, mpfr_get_prec (op));

  printf ("%ld %ld\n", mpfr_get_emin (), mpfr_get_emax ());
  mpfr_set_emin (em1);
  mpfr_set_emax (em2);
  printf ("%ld %ld\n", mpfr_get_emin (), mpfr_get_emax ());

  printf ("%ld %ld %ld %ld\n", mpfr_get_emin_min (), mpfr_get_emin_max (), mpfr_get_emax_min (), mpfr_get_emax_max ());

  mpfr_set (rop, op, MPFR_RNDN);
  mpfr_check_range (rop, t, MPFR_RNDN);
  mpfr_printf ("%.Re\n", rop);

  mpfr_set (rop, op, MPFR_RNDN);
  mpfr_subnormalize (rop, t, MPFR_RNDN);
  mpfr_printf ("%.Re\n", rop);

  flags ();

  mpfr_set_underflow ();
  mpfr_set_overflow ();
  mpfr_set_divby0 ();
  mpfr_set_nanflag ();
  mpfr_set_inexflag ();
  mpfr_set_erangeflag ();

  flags ();

  mpfr_clear_underflow ();
  mpfr_clear_overflow ();
  mpfr_clear_divby0 ();
  mpfr_clear_nanflag ();
  mpfr_clear_inexflag ();
  mpfr_clear_erangeflag ();

  flags ();

  mpfr_flags_set(MPFR_FLAGS_NAN|MPFR_FLAGS_ERANGE|MPFR_FLAGS_DIVBY0);
  flags ();
  int saved_flags = mpfr_flags_save();
  int test_all_flags = mpfr_flags_test(MPFR_FLAGS_ALL);
  int test_one_flag = mpfr_flags_test(MPFR_FLAGS_ERANGE);
  printf ("%d %d\n", test_all_flags, test_one_flag);
  mpfr_flags_clear(MPFR_FLAGS_ALL);
  flags ();
  mpfr_flags_restore(saved_flags, MPFR_FLAGS_ERANGE|MPFR_FLAGS_DIVBY0|MPFR_FLAGS_OVERFLOW);
  flags ();

  mpfr_clear (rop);
}

int main ()
{
  mpfr_t op;
  mpfr_init (op);
  mpfr_exp_t e1, e2;
  int t;

  e1 = -123;
  e2 = 123;
  t = mpfr_set_d (op, 1. / 3, MPFR_RNDN);

  all(op, t, e1, e2);
  printf("\n");

  mpfr_clear (op);
  return 0;
}
